path_subkeys.go

```diff
@@ -0,0 +1,190 @@
+// provide all subkeys with nesting fully intact. The modifications are made
+// to the input in-place. maxDepth will denote how deep to traverse. A maxDepth
+// of 0 is the equivalent of no limit.
+func removeValues(input map[string]interface{}, maxDepth int) {
+	var walk func(interface{}, int)
+
+	walk = func(in interface{}, depth int) {
+		val := reflect.ValueOf(in)
+
+		if val.Kind() == reflect.Map {
+			for _, k := range val.MapKeys() {
+				v := val.MapIndex(k)
+				m := in.(map[string]interface{})
+
+				switch t := v.Interface().(type) {
+				case map[string]interface{}:
+					// Only continue walking if we have not reached max depth
+					// and the underlying map has at least 1 key. The key is
+					// otherwise treated as a leaf node and thus set to nil.
+					// Setting to nil if the max depth is reached is crucial in
+					// that it prevents leaking secret data as the input map is
+					// being modified in-place
+					if currentDepth := depth + 1; (maxDepth == 0 || currentDepth <= maxDepth) && len(t) > 0 {
+						walk(t, currentDepth)
+					} else {
+						m[k.String()] = nil
+					}
+				default:
+					m[k.String()] = nil
+				}
+			}
+		}
+	}
+
+	walk(input, 1)
+}
+
+// pathSubkeysRead handles ReadOperation requests for a specified path. Subkeys
+// that exist within the entry specified by the provided path will be retrieved.
+// This is done by stripping the secret data by replacing all underlying values of
+// leaf keys with null.
+func (b *versionedKVBackend) pathSubkeysRead() framework.OperationFunc {
+	return func(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+		key := data.Get("path").(string)
+		depth := data.Get("depth").(int)
+
+		lock := locksutil.LockForKey(b.locks, key)
+		lock.RLock()
+		defer lock.RUnlock()
+
+		meta, err := b.getKeyMetadata(ctx, req.Storage, key)
+		if err != nil {
+			return nil, err
+		}
+		if meta == nil {
+			return nil, nil
+		}
+
+		versionNum := meta.CurrentVersion
+		versionParam := data.Get("version").(int)
+
+		if versionParam > 0 {
+			versionNum = uint64(versionParam)
+		}
+
+		versionMetadata := meta.Versions[versionNum]
+		if versionMetadata == nil {
+			return nil, nil
+		}
+
+		resp := &logical.Response{
+			Data: map[string]interface{}{
+				"subkeys": nil,
+				"metadata": map[string]interface{}{
+					"version":         versionNum,
+					"created_time":    ptypesTimestampToString(versionMetadata.CreatedTime),
+					"deletion_time":   ptypesTimestampToString(versionMetadata.DeletionTime),
+					"destroyed":       versionMetadata.Destroyed,
+					"custom_metadata": meta.CustomMetadata,
+				},
+			},
+		}
+
+		if versionMetadata.DeletionTime != nil {
+			deletionTime, err := ptypes.Timestamp(versionMetadata.DeletionTime)
+			if err != nil {
+				return nil, err
+			}
+
+			if deletionTime.Before(time.Now()) {
+				return logical.RespondWithStatusCode(resp, req, http.StatusNotFound)
```

Comment:
Do we want to return the response here? What about returning `logical.ErrorResponse` instead?

Reply:
`logical.RespondWithStatusCode` is a [helper function](https://github.com/hashicorp/vault/blob/13b8306575e592086bbcc05c7787b568de1146ff/sdk/logical/response.go#L171) that takes a `logical.Response`, the incoming `logical.Request` and an HTTP status code. It will add `HTTPStatusCode` to the `Data` field in the `logical.Response`. Vault's response handling logic will pull this out and use it as the response status. This is a means to ensure that a 404 response is provided if the secret entry has been deleted (we also check if it is destroyed). This check is consistent with some of the other KV endpoints. See the read for `data/:path`:

https://github.com/hashicorp/vault-plugin-secrets-kv/blob/9f2cb0e9e74d401758451f3706353df72c899cbe/path_data.go#L126-L143